File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.java
package ar.com.wolox.android.example.ui.viewpager.request;

import ar.com.wolox.android.example.model.Post;
import ar.com.wolox.android.example.network.PostService;
import ar.com.wolox.wolmo.core.presenter.BasePresenter;
import ar.com.wolox.wolmo.networking.retrofit.RetrofitServices;
import ar.com.wolox.wolmo.networking.retrofit.callback.NetworkCallback;

import javax.inject.Inject;

import okhttp3.ResponseBody;

public class RequestPresenter extends BasePresenter<IRequestView> {

    private static final int POST_ID = 1;

    @Inject RetrofitServices mRetrofitServices;

    @Inject
    RequestPresenter() {}

    @Override
    public void onViewAttached() {
        mRetrofitServices.getService(PostService.class).getPostById(POST_ID).enqueue(new NetworkCallback<Post>() {
            @Override
            public void onResponseSuccessful(final Post response) {
                runIfViewAttached(new Consumer<IRequestView>() {
                    @Override
                    public void accept(IRequestView view) {
                        view.setNewsTitle(response.getTitle());
                        view.setNewsBody(response.getBody());
                    }
                });
            }

            @Override
            public void onResponseFailed(ResponseBody responseBody, int code) {
                runIfViewAttached(new Runnable() {
                    @Override
                    public void run() {
                        getView().showError();
                    }
                });
            }

            @Override
            public void onCallFailure(Throwable t) {
                runIfViewAttached(new Runnable() {
                    @Override
                    public void run() {
                        getView().showError();
                    }
                });
            }
        });
    }
}

File: app/src/main/java/ar/com/wolox/android/example/ui/viewpager/request/RequestPresenter.kt
package ar.com.wolox.android.example.ui.viewpager.request

import ar.com.wolox.android.example.model.Post
import ar.com.wolox.android.example.network.builder.networkRequest
import ar.com.wolox.android.example.network.repository.PostRepository
import ar.com.wolox.wolmo.core.presenter.CoroutineBasePresenter
import kotlinx.coroutines.launch
import javax.inject.Inject

class RequestPresenter @Inject constructor(
    private val postRepository: PostRepository
) : CoroutineBasePresenter<RequestView>() {

    fun onSearchRequested(postId: Int?) = launch {
        if (postId == null) {
            view?.showInvalidInput()
            return@launch
        }

        networkRequest(postRepository.getPostById(postId)) {
            onResponseSuccessful { response -> showPost(response!!) }
            onResponseFailed { _, _ -> view?.showError() }
            onCallFailure { view?.showError() }
        }
    }

    private fun showPost(post: Post) = view?.run {
        setTitle(post.title)
        setBody(post.body)
    }
}
